feat: add Milvus vectorDB

zc277584121 opened the pull request:
# What does this PR do?
feat: add [Milvus](https://milvus.io/) vectorDB

note: I use the MilvusClient to implement it instead of AsyncMilvusClient, because when I tested AsyncMilvusClient, it would raise issues about evenloop, which I think AsyncMilvusClient SDK is not robust enough to be compatible with llama_stack framework.

## Test Plan
have passed the unit test and ene2end test
Here is my end2end test logs, including the client code, client log, server logs from inline and remote settings
[test_end2end_logs.zip](https://github.com/user-attachments/files/18964391/test_end2end_logs.zip)

terrytangyuan reviewed (commented):
Thank you! @franciscojavierarceo given your prior experience on this, would you like to help review this PR?

franciscojavierarceo commented:
yeah of course!

franciscojavierarceo commented on docs/source/concepts/index.md:
```diff
@@ -25,7 +25,7 @@ We are working on adding a few more APIs to complete the application lifecycle.
 
 The goal of Llama Stack is to build an ecosystem where users can easily swap out different implementations for the same API. Examples for these include:
 - LLM inference providers (e.g., Fireworks, Together, AWS Bedrock, Groq, Cerebras, SambaNova, etc.),
-- Vector databases (e.g., ChromaDB, Weaviate, Qdrant, FAISS, PGVector, etc.),
```
please also update the docs section as well under https://github.com/meta-llama/llama-stack/tree/main/docs/source/providers/vector_io/

See this PR as a reference https://github.com/meta-llama/llama-stack/pull/1195

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/milvus/__init__.py:
```diff
@@ -0,0 +1,21 @@
+
+
+async def get_adapter_impl(config: MilvusVectorIOConfig, deps: Dict[Api, ProviderSpec]):
+    from .milvus import MilvusVectorIOAdapter
```
I *think* the inline and remote implementations are equivalent.

If that is, indeed, the case can you follow the pattern that's done by Chroma? 

See here: https://github.com/meta-llama/llama-stack/blob/main/llama_stack/providers/inline/vector_io/chroma/__init__.py#L15

In short, they just import the inline implementation in the `__init__.py` and don't have an inline `chromadb.py`.

franciscojavierarceo commented on llama_stack/providers/inline/vector_io/milvus/milvus.py:
```diff
@@ -0,0 +1,143 @@
+
+        data = []
+        for i, (chunk, embedding) in enumerate(zip(chunks, embeddings, strict=False)):
+            chunk_id = f"{chunk.metadata['document_id']}:chunk-{i}"
```
It's probably better to generate the chunk ID from the text and the document ID in case the order of insertion changes. See this function: https://github.com/meta-llama/llama-stack/blob/main/llama_stack/providers/inline/vector_io/sqlite_vec/sqlite_vec.py#L234

Maybe that can be moved to a utils folder.

franciscojavierarceo commented on llama_stack/providers/inline/vector_io/milvus/milvus.py:
```diff
@@ -0,0 +1,143 @@
+                    "chunk_content": chunk.model_dump(),
+                }
+            )
+        self.client.insert(
```
Please have some error handling and log the collection name if insertion fails

franciscojavierarceo reviewed (changes requested):
Thank you for this! Very excited to see Milvus. Left some initial feedback, could you make some changes? 

Also, could you update your PR description with the output of your test run?

franciscojavierarceo commented on llama_stack/providers/inline/vector_io/milvus/milvus.py:
```diff
@@ -0,0 +1,143 @@
+            f"Chunk length {len(chunks)} does not match embedding length {len(embeddings)}"
+        )
+        if not self.client.has_collection(self.collection_name):
+            self.client.create_collection(self.collection_name, dimension=len(embeddings[0]), auto_id=True)
```
I know bounded staleness is the the OOTB consistency configuration. It may be useful to make that explicit. Alternatively, I could see strong consistency being a useful default.

zc277584121 commented:
sure, I will fix them soon

zc277584121 commented:
@franciscojavierarceo I have updated this PR to solve all the problems you mentioned, please check them again, thanks.

ehhuang commented on llama_stack/providers/remote/vector_io/milvus/milvus.py:
```diff
@@ -0,0 +1,165 @@
+            self.client = MilvusClient(**self.config.model_dump(exclude_none=True))
+        else:
+            logger.info(f"Connecting to Milvus Lite at: {self.config.db_path}")
+            uri = self.config.model_dump(exclude_none=True)["db_path"]
```
is this equivalent to `uri = self.config.db_path`?

zc277584121 replied:
yes, i will fix it

ehhuang commented on llama_stack/providers/inline/vector_io/milvus/config.py:
```diff
@@ -0,0 +1,20 @@
+
+@json_schema_type
+class MilvusVectorIOConfig(BaseModel):
+    db_path: str
```
Milvus n00b: how would users get this db_path?

ehhuang commented on tests/client-sdk/vector_io/test_vector_io.py:
```diff
@@ -12,6 +12,7 @@
     "faiss",
     # TODO: add sqlite_vec to templates
     # "sqlite_vec",
+    # "milvus",
```
Why is this commented out?

franciscojavierarceo replied:
sqlite-vec was commented out by @yanxi0830 and I suppose this just followed suit. 

Would be good to uncomment both to test both.

zc277584121 replied:
Yes, I just followed it. In the new update, I uncommeted it

ehhuang commented:
Thanks for working on this! Maybe I missed it, but where are the unit tests added?

Could we also add some simple tests that test the db querying? Maybe add some dummy texts to the db and make sure the most relevant chunk is at the top of the result.

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/milvus/milvus.py:
```diff
@@ -0,0 +1,165 @@
+            raise ValueError(f"Vector DB {vector_db_id} not found")
+
+        return await index.query_chunks(query, params)
+
```
```suggestion
TODO: refactor this generate_chunk_id along with the `sqlite-vec` implementation into a separate utils file
```

franciscojavierarceo replied:
Can you add this comment as a follow up?

zc277584121 replied:
sure

zc277584121 commented:
@franciscojavierarceo @ehhuang Thank you for the review, and I have updated the code again: uncomment the test case, add todo in the tail, removed the use of `model_dump`

zc277584121 commented:
@franciscojavierarceo sure, I am work on this

franciscojavierarceo commented on tests/client-sdk/vector_io/test_vector_io.py:
```diff
@@ -12,6 +12,7 @@
     # TODO: add sqlite_vec to templates
     # "sqlite_vec",
+    "milvus",
 ]
```
+1

You could take a look at this example: https://github.com/meta-llama/llama-stack/blob/66cd128ab51aff0b649c8ae59d7ec139a54913c1/llama_stack/providers/tests/vector_io/test_sqlite_vec.py#L24

zc277584121 replied:
ok

zc277584121 commented:
@ehhuang @franciscojavierarceo have updated the unittest, I found llama_stack/providers/tests/vector_io/test_vector_io.py can test all providers, so I add milvus to the conftest.py and fixtures.py. Then run the unittest, and it is my output:

```log
============================= test session starts ==============================
collecting ... collected 24 items / 21 deselected / 3 selected

llama_stack/providers/tests/vector_io/test_vector_io.py::TestVectorIO::test_banks_list[-milvus] 
llama_stack/providers/tests/vector_io/test_vector_io.py::TestVectorIO::test_banks_register[-milvus] 
llama_stack/providers/tests/vector_io/test_vector_io.py::TestVectorIO::test_query_documents[-milvus] PASSEDPASSEDThe scores are: [0.25088539719581604, 0.2342979907989502, 0.22884511947631836]
PASSED

========= 3 passed, 21 deselected, 27843 warnings in 333.04s (0:05:33) =========
```

franciscojavierarceo commented:
@zc277584121 there were some updates made to the tests, so good to be aware of them: https://github.com/meta-llama/llama-stack/pull/1398

ehhuang commented on llama_stack/providers/remote/vector_io/milvus/milvus.py:
```diff
@@ -0,0 +1,166 @@
+    hash_input = f"{document_id}:{chunk_text}".encode("utf-8")
+    return str(uuid.UUID(hashlib.md5(hash_input).hexdigest()))
+
+# TODO: refactor this generate_chunk_id along with the `sqlite-vec` implementation into a separate utils file
```
do you mind filing an issue as well?

zc277584121 replied:
Sure, after this PR is merged, I will open an issue about it.

ehhuang reviewed (approved):
Thanks for following up on the tests!

zc277584121 commented:
@franciscojavierarceo I noticed from https://github.com/meta-llama/llama-stack/pull/1398 that the change in config.py from `sample_config` to `sample_run_config`, so I did the same as follow in the new commit `34194ff`. I also found `__distro_dir__` was added into the `config.py` of sqlite_vec and faiss, but not in other vectordb providers, so I keep it unchanged.  In this situation, the unittest would failed with this error:
```
test_vector_io.py::test_vector_db_retrieve[txt=:vis=:emb=all-miniLM-L6-V2] ERROR [ 14%]
test setup failed
../conftest.py:185: in llama_stack_client
    config = distro_from_adhoc_config_spec(config)
../conftest.py:157: in distro_from_adhoc_config_spec
    provider_config = replace_env_vars(provider_config_type.sample_run_config(__distro_dir__=distro_dir))
E   TypeError: MilvusVectorIOConfig.sample_run_config() got an unexpected keyword argument '__distro_dir__'
```
So I think it would also fail in other vectordb provider's cases. I don't know if this problem will be solved in the future, or there is any other way to run the unittest. I think I have tried my best to do this.

But the end2end test still works well.

franciscojavierarceo commented:
FYI @ashwinb

franciscojavierarceo commented:
> @franciscojavierarceo I noticed from #1398 that the change in config.py from `sample_config` to `sample_run_config`, so I did the same as follow in the new commit `34194ff`. I also found `__distro_dir__` was added into the `config.py` of sqlite_vec and faiss, but not in other vectordb providers, so I keep it unchanged. In this situation, the unittest would failed with this error:
> 
> ```
> test_vector_io.py::test_vector_db_retrieve[txt=:vis=:emb=all-miniLM-L6-V2] ERROR [ 14%]
> test setup failed
> ../conftest.py:185: in llama_stack_client
>     config = distro_from_adhoc_config_spec(config)
> ../conftest.py:157: in distro_from_adhoc_config_spec
>     provider_config = replace_env_vars(provider_config_type.sample_run_config(__distro_dir__=distro_dir))
> E   TypeError: MilvusVectorIOConfig.sample_run_config() got an unexpected keyword argument '__distro_dir__'
> ```
> 
> So I think it would also fail in other vectordb provider's cases. I don't know if this problem will be solved in the future, or there is any other way to run the unittest. I think I have tried my best to do this.
> 
> But the end2end test still works well.

Hmmm, that doesn't sound ideal. I'll defer to @ashwinb there but since @ehhuang approved, I won't block.

Probably though I'd try to resolve it if possible.

ashwinb commented:
Hm I think I did something weird here after trying to fix errors when testing locally.

ashwinb commented:
Copied this PR into https://github.com/meta-llama/llama-stack/pull/1467 and fixed forwards. Sorry about this @zc277584121
